### ISSUE TITLE
.Net 6

### ISSUE THREAD

**Issue by TheSorm** [Tewr/BlazorWorker#65]
Is .Net6 supported yet? Or are there work arounds to get it to work?

I used the standart example from this git:
```

var worker = await workerFactory.CreateAsync();

var service = await worker.CreateBackgroundServiceAsync<TestService>();

var localParameterValue = 42;
var result = await service.RunAsync(s => s.MyMethod(localParameterValue));
```

```
public class TestService
{
    public int MyMethod(int parameter)
    {
        int i = 0;
        while (i < 5000000) i += (i * parameter);
        return i;
    }
}
```

 and got:
`/__w/1/s/src/mono/mono/metadata/icall.c:6639 <disabled> dotnet.6.0.0-preview.7.21377.19.js:1:17346`
```
WASM-WORKER: undefined 65ddd4df-7800-4bcd-87e3-c2164e3ffdda:70:21
    printErr blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:70
    abort https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    _abort https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1348656
    <anonym> https://localhost:5001/_framework/dotnet.wasm:517767
    <anonym> https://localhost:5001/_framework/dotnet.wasm:420446
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1478160
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1441913
    <anonym> https://localhost:5001/_framework/dotnet.wasm:707208
    <anonym> https://localhost:5001/_framework/dotnet.wasm:88413
    <anonym> https://localhost:5001/_framework/dotnet.wasm:41808
    <anonym> https://localhost:5001/_framework/dotnet.wasm:284802
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1245660
    <anonym> https://localhost:5001/_framework/dotnet.wasm:468414
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1478160
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1441913
    <anonym> https://localhost:5001/_framework/dotnet.wasm:707208
    <anonym> https://localhost:5001/_framework/dotnet.wasm:88413
    <anonym> https://localhost:5001/_framework/dotnet.wasm:41808
    <anonym> https://localhost:5001/_framework/dotnet.wasm:284802
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1245660
    <anonym> https://localhost:5001/_framework/dotnet.wasm:468414
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1478160
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1441913
    <anonym> https://localhost:5001/_framework/dotnet.wasm:707208
    <anonym> https://localhost:5001/_framework/dotnet.wasm:88413
    <anonym> https://localhost:5001/_framework/dotnet.wasm:41808
    <anonym> https://localhost:5001/_framework/dotnet.wasm:447521
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1478160
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1441913
    <anonym> https://localhost:5001/_framework/dotnet.wasm:707208
    <anonym> https://localhost:5001/_framework/dotnet.wasm:88413
    <anonym> https://localhost:5001/_framework/dotnet.wasm:1428159
    _mono_wasm_invoke_method https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    _call_method_with_converted_args https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    call_method https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    call_static_method https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    onReady blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:26
    onmessage blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:134
    callRuntimeCallbacks https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    postRun https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    doRun https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    run https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    runCaller https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    removeRunDependency https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    onmessage blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:114
```
```
Init method [BlazorWorker.WorkerBackgroundService]BlazorWorker.WorkerBackgroundService.WorkerInstanceManager:Init failed ReferenceError: showErrorNotification is not defined 65ddd4df-7800-4bcd-87e3-c2164e3ffdda:28:25
    onReady blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:28
    onmessage blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:134
    callRuntimeCallbacks https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    postRun https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    doRun https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    run https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    runCaller https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    removeRunDependency https://localhost:5001/_framework/dotnet.6.0.0-preview.7.21377.19.js:1
    onmessage blob:https://localhost:5001/65ddd4df-7800-4bcd-87e3-c2164e3ffdda:114
```
`ReferenceError: showErrorNotification is not defined 65ddd4df-7800-4bcd-87e3-c2164e3ffdda:71:13`

**Comment by Tewr** [comment]
Hello @Sen-Gupta. I didn't take the time to look into this yet. The error is thrown here: https://github.com/Tewr/BlazorWorker/blob/69dcb9b42193dfeb15aa138ba4411e652abd9efd/src/BlazorWorker/BlazorWorker.js#L86 and I guess this method no longer exists. But it's probably a detail. We are inside the mono `printErr` method, and `line` seems to be undefined, so its probably part of some other error and the current stack trace does not really help us.

The Blazor/mono bootstrapper has been rewritten and the BlazorWorker bootstrapper is no longer compatible. Hard to say if this will result in some small quick fix or a complete rewrite of the BlazorWorker bootstrapper.

When I will have some time to look into this, my first step after reproducing the error will be a side-by-side comparison/analysis of the Blazor bootstrap code of dotnet 6 with the BlazorWorker bootstrapper, eventually comparing the bootstrapper of Blazor on dotnet5 vs the dotnet 6 one.

Some relevant reads will be
https://github.com/dotnet/aspnetcore/blob/release/6.0-rc2/src/Components/Web.JS/src/Boot.WebAssembly.ts
and
https://github.com/dotnet/aspnetcore/blob/release/6.0-rc2/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

**Comment by Tewr** [comment]
So I managed to find some time for this, and looks like its "simply" a missing dll in the core bootloader  here:  
https://github.com/Tewr/BlazorWorker/blob/69dcb9b42193dfeb15aa138ba4411e652abd9efd/src/BlazorWorker/WorkerProxyDependencies.cs#L18-L68

Probably some important dll as we won't even manage to get the info in an error message. I managed to identify this being the problem by bruteforcing and loading the full blazor framework into the worker.

Now I'd like to reduce this list of course. A first beta version could be just to exlude the obvious blazor framework dlls.

The best solution to this problem would be not to deal with dlls at all, as discussed in #23. I may be lazy enough (I'm pretty sure this is not the last time I'll have to walk the dependency list more or less manually) to finally try to implement something. It would have to be a compile-time component, but maybe a Roslyn Code generator could help me find the info I need (a map of assembly name -> filename)

**Comment by Tewr** [comment]
My previous findings was due to a lack of coffee I think. In any case it was all false conclusions

**Comment by htapal** [comment]
@Tewr can you explain your last comment? Is this working in .NET 6 now that it's released?

**Comment by Tewr** [comment]
I've spent a few hours on different hypothesises, but no luck yet. The differences in the new bootloader code seems subtle at a first glance, unfortunately the error messages and browser crashes I get at the moment are not very helpful. There is a new resource loader class that I have to figure out the inner workings of as well, maybe there is some detail I'm missing in there.

**Comment by Tewr** [comment]
So finally some good news. I've managed to get it working, basically by carefully comparing my bootloader with the new one from .net6, changing line by line, and finally kind of beating it with a hammer whenever I had browser crashes and cryptic error messages (trying to remove, add, and undo modifs compared to mine). 

My conclusions to why I got it working are quite vague, but the major changes are using the object `BINDING` instead of `MONO` for static method binding, and NOT calling the "new" method  `MONO.mono_wasm_runtime_ready();` like they do here: https://github.com/dotnet/aspnetcore/blob/2d82d2cc6338b6e720da3d7cd77df680c0bdbf5f/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts#L451

Anyway, I will be preparing at least a pre-release package in a few days. For anyone in a hurry, the working version is in this pull request: https://github.com/Tewr/BlazorWorker/pull/66
